File: 042_Trapping_Rain_Water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        stack, res = [], 0
        for i in range(len(height)):
            while stack and height[i] > height[stack[-1]]:
                top = stack.pop()
                if not stack:
                    break
                d = i - stack[-1] - 1
                h = min(height[i], height[stack[-1]]) - height[top]
                res += d * h
            stack.append(i)
        return res


class Solution:
    def trap(self, height: List[int]) -> int:
        left, right, left_max, right_max, res = 0, len(height) - 1, 0, 0, 0
        while left < right:
            if height[left] < height[right]:
                if height[left] > left_max:
                    left_max = height[left]
                else:
                    res += max(left_max - height[left], 0)
                left += 1
            else:
                if height[right] > right_max:
                    right_max = height[right]
                else:
                    res += max(right_max - height[right], 0)
                right -= 1
        return res

File: 042_Trapping_Rain_Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() < 3) return 0;
        int left = 0, right = height.size() - 1;
        int left_max = height[left], right_max = height[right], res = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] > left_max) left_max = height[left];
                else res += left_max - height[left];
                ++left;
            }
            else {
                if (height[right] > right_max) right_max = height[right];
                else res += right_max - height[right];
                --right;
            }
        }
        return res;
    }
};
